### DISCUSSION TITLE
SmartFormComponent component work correctly in dev, when i build production build that time is not working

### DISCUSSION THREAD

**Discussion by hirenbhut93** [react-hook-form/react-hook-form#2362] [Q&A]
https://react-hook-form.com/advanced-usage#SmartFormComponent

Means `register` arg is not available in Input or Select, those pass via Form component but only received arg name and options in production build,

Work correctly when i develop using `npm start`, but issue when i build production using `npm run build`

**Comment by bluebill1049** [comment]
need more detail and code to reproduce such issue

**Reply by bluebill1049** [reply]
we using this pattern and prod and there are no issues. you need to do some debug when `register` is not available.

**Reply by hirenbhut93** [reply]
@bluebill1049  I have already debug using console.log(register), then run using `npm start` it's working but in production build return `register`  is `undefined`

Any version or production isssue?

**Reply by hirenbhut93** [reply]
Any alternative for debug for production

**Reply by bluebill1049** [reply]
debug on production and see what's going on there.

**Reply by bluebill1049** [reply]
especially around context and props drilling.

**Reply by hirenbhut93** [reply]
@bluebill1049 I have change the some code for export form like

```
export function Form({
  defaultValues,
  children,
  schema = null,
  onSubmit,
  ...props
}) {
  const methods = useForm({
    defaultValues,
    resolver: schema ? yupResolver(schema) : null,
  });
  const { handleSubmit } = methods;

  function renderByType(child) {
    if (child.type.name === "Field" || child.type.name === "FieldCheckbox") {
      return React.createElement(child.type, {
        ...{
          ...child.props,
          key: child.props.name,
          label: child.props.label ?? labelFromSchema(schema, child.props.name),
          register: methods.register,
          error: deepFind(methods.errors, child.props.name),
          isRequired: isRequiredField(schema, child.props.name),
        },
      });
    } else if (child.type.name === "SubmitButton") {
      return React.createElement(child.type, {
        ...{
          ...child.props,
          formState: methods.formState,
          key: "btn-submit",
        },
      });
    }

    return child;
  }

  return (
    <form
      onSubmit={handleSubmit((data, e) => onSubmit(data, methods.setError, e))}
      {...props}
    >
      {Array.isArray(children)
        ? children.map((child) => renderByType(child))
        : children}
    </form>
  );
}
```

any issue with this code?

**Reply by bluebill1049** [reply]
looks fine to me, i think you may need to do some debug with react dev tool with your production code. see what's going wrong.

**Reply by hirenbhut93** [reply]
@bluebill1049  I have prod build using `npm run build -- --profile` so is work correctly/perfectly. So i don't understanding what is magic with me.

Also check this is my form code

```
import React from "react";
import { Modal } from "react-bootstrap";
import * as yup from "yup";
import {
  Field,
  FieldCheckbox,
  Form,
  SubmitButton,
} from "../../../../components/Bootstrap/Form";
import httpService from "../../../../services/http.service";
import { formResponse } from "../../../../utils/common";

export default function FormUrl({ handleClose, subId, secItem = {} }) {
  const isEdit = secItem !== null;

  const schema = yup.object().shape({
    name: yup.string().required().label("Name"),
    summary: yup.string().label("Summary"),
    visible: yup.boolean().label("Visible"),
  });

  const onSubmit = async (data, setError) => {
    const url = isEdit
      ? `section/edit?id=${secItem.id}`
      : `lsection/add?sid=${subId}`;
    const res = isEdit
      ? await httpService.put(url, data)
      : await httpService.post(url, data);
    if (formResponse(res, setError)) {
      handleClose({
        from: "section",
        isNew: !isEdit,
        secItem: res.data,
      });
    }
  };

  return (
    <Modal show={true} onHide={handleClose} backdrop="static" keyboard={false}>
      <Modal.Header closeButton>
        <h5 className="modal-title">
          Section
        </h5>
      </Modal.Header>
      <Modal.Body>
        <Form
          schema={schema}
          onSubmit={onSubmit}
          defaultValues={secItem ?? { visible: 1 }}
        >
          <Field type="text" name="name" />
          <Field type="textarea" name="summary" />
          <FieldCheckbox name="visible" />
          <SubmitButton label="Save" className="btn btn-block btn-success" />
        </Form>
      </Modal.Body>
    </Modal>
  );
}
```

**Reply by bluebill1049** [reply]
shouldn't useForm sit inside Form component? instead outside?

**Reply by hirenbhut93** [reply]
> shouldn't useForm sit inside Form component? instead outside?

@bluebill1049 , Sorry I dont't understand what you told me, please give me details what change is try to do.

https://github.com/react-hook-form/react-hook-form/discussions/2362#discussioncomment-42320 this is my form component

**Reply by bluebill1049** [reply]
oh oops, sorry did a search and it jumps to the above comment...

**Reply by bluebill1049** [reply]
can you turn this into a codesandbox? if this is an issue. alternative can you share your prod build URL? happy to take a look there too.

**Reply by hirenbhut93** [reply]
@bluebill1049 Please check codesand box https://codesandbox.io/s/upbeat-hofstadter-749ci?file=/src/App.js and github project https://github.com/hirenbhut93/hook-form-issue

This is work in devlopment but not work production build

**Reply by bluebill1049** [reply]
please share a prod site link, too lazy to clone repo and dig around. share the prod link for the form, i will take a look.

**Reply by hirenbhut93** [reply]
@bluebill1049  please check https://hook-form-issue.vercel.app/ this is production

and this is development  https://codesandbox.io/s/upbeat-hofstadter-749ci?file=/src/App.js

**Reply by bluebill1049** [reply]
looks like nothing is register in prod build.

**Reply by hirenbhut93** [reply]
@bluebill1049 means?

I request to you please clone this project https://github.com/hirenbhut93/hook-form-issue and run `npm start` and 'npm run build' then check

**Reply by bluebill1049** [reply]
couple things: 
<img width="766" alt="Screen Shot 2020-07-27 at 8 48 07 pm" src="https://user-images.githubusercontent.com/10513364/88533988-83771c00-d04a-11ea-869b-e92ebf7973d7.png">
it's suspicious with inline function like above.

also, I would suggest following what I had in the doc: https://react-hook-form.com/advanced-usage#SmartFormComponent because that's what we use at work in Prod, it's working.

**Reply by bluebill1049** [reply]
why you request me to do something? I don't get it, mate, I am here to offer free help... I am done here.

**Reply by hirenbhut93** [reply]
> looks like nothing is register in prod build.
@bluebill1049 
Sorry I don't understand what is told me. can need build director?

**Reply by hirenbhut93** [reply]
> why you request me to do something? I don't get it, mate, I am here to offer free help... I am done here.

@bluebill1049 you are right, and Sorry for that

but i beginner of react, so i don't know how to  work actually.
